### ISSUE TITLE
do is arguably not redundant when used for non-decreasing indentation.

### ISSUE THREAD

**Issue by galenhuntington** [ndmitchell/hlint#978]
Sorry, another issue.  The switch `-XNondecreasingIndentation` is [on by default](https://downloads.haskell.org/~ghc/8.10.1/docs/html/users_guide/bugs.html#context-free-syntax) and can be utilized to keep indentation under control.  But consider these uses:

```haskell
main = do
   main2
   print =<< do
   fmap reverse getContents
main2 = do
   print "x"
   if False then main else do
   return ()
```

hlint outputs

```
test.hs:3:14-15: Warning: Redundant do
Found:
  do
Perhaps you should remove it.

test.hs:7:28-29: Warning: Redundant do
Found:
  do
Perhaps you should remove it.

2 hints
```

Arguably, these aren't redundant because they serve to tame indentation, but even if so, the advice to remove them should probably be coupled with a notice to re-indent.

Removing `do` (without re-indenting) in those cases is simply a syntax error, but with `-XBlockArguments`, it could be more insidious.  Example:

```haskell
{-# LANGUAGE BlockArguments #-}
import Data.Char
import Text.Printf
foo :: String
foo = do
   "2x"
   map toUpper $ printf "%s" do
   "a"
```

If the `do` is removed, not only is the syntax still valid, but the result will even type-check and compile, and only fail at runtime.

I tried both of these with `--refactor`, and in neither case did it re-indent.  So I think it's fair to call it a bug.

**Comment by ndmitchell** [comment]
Thanks for the report. Sadly, this style of do is also very fiddly to detect... I guess keep track of the nearest encoding do and see if that's got the same indent level? Is that sufficient? Is that complete?

**Comment by galenhuntington** [comment]
Yeah I think that would suffice (comparing the indentation level to the enclosing `do`).

**Comment by ndmitchell** [comment]
Seems to work nicely. Fix pushed.
